### DISCUSSION TITLE
`v12.2` [Middleware GA] Alternative for removed `page` from `NextRequest` (breaking authentication use case)

### DISCUSSION THREAD

**Discussion by bennettdams** [vercel/next.js#38296] [Ideas]
## Use case

I would like to use middleware for authentication. Every page should only be available to authenticated users.
I can check the authentication status based on cookies on the request.
If the user is not signed in, I can redirect him to a sign in page. But I only want/need to do this check for real pages/routes, not for everything else.

```ts
if (
  !req.page.name ||
  !!isRevalidationRequest ||
  req.page.name === ROUTES.SIGN_IN ||
  req.page.name.startsWith("/api")
) {
  // the request is not for one of our pages or it is for the sign in page or it is an API route request, go on
  return NextResponse.next();
} else {
  // check auth and redirect if necessary...
}
```

## Background

With `v12.2` the `page` field does not exist anymore on the `NextRequest` in middleware.
We used `page` to find out whether a request in the middleware was for a page route, so we can check the authentication status and redirect to a sign in page if necessary.

This was already [discssued in the middleware RFC](https://github.com/vercel/next.js/discussions/29750#discussioncomment-2122618), but without an official response.

I know that `page` was never officially documented, but now that middleware is in "general availability", I feel like there is a gap for this functionality, hence this feature request.

## Issue

Now that [**all** requests are intercepted by middleware](https://nextjs.org/docs/messages/middleware-upgrade-guide#executing-middleware-on-internal-nextjs-requests) and we **also** [don't have `page` anymore](https://nextjs.org/docs/messages/middleware-upgrade-guide#no-more-page-match-data), there's no good way to identify a request that needs to be authenticated anymore. This is because middleware now also runs for e.g.
 - `_next`/internal requests 
 - public assets (like `favicon.ico`)
 - fetches to custom URLs (e.g. an external backend like `/my-backend/some-data`) when using a proxy/rewrites

### Describe the solution you'd like

I would like to be able to identify in middleware whether a request is for a page or not.
Maybe Next.js could provide a helper function that we can important in our code to detect a page, which is based on the internal routing our app has. 

### Describe alternatives you've considered

The only thing that comes to my mind is writing down all routes in the project and match the URL in the middleware, but this is a brittle solution, as a new route can easily be forgotten in this setup. 
Maybe the list of registered routes can be provided via a public API of Next.js' internal routing, so there's no way to forget a route.

There's [another proposed solution](https://github.com/vercel/next.js/discussions/38159#discussioncomment-3056398), but it also needs valid page routes to be defined upfront and additionally is using Next.js internals, which is not a good idea as well imo.

## Update: "Workaround"

Looking at the [official authentication example](https://github.com/vercel/examples/blob/main/edge-functions/jwt-authentication/middleware.ts) does not help, as they only have specific routes that are protected (which can easily done with the new matcher config).
What I have found [is another example](https://github.com/vercel/next.js/pull/38419/files) (for i18n routing), where the same check (_Is the request for a real page?_) has to be done. Here, there are checks in place:

1. Is this an internal Next.js request?
2. Is this an API request?
3. Is this a request for a file from the public folder?

In my case, I have to add three more checks:

4. Is this a request for my custom API?
5. ~~Is this a request coming from on-demand revalidation?~~ Update: [this was changed in `v12.2.0`](https://github.com/vercel/next.js/pull/38562).
6. Is this a request for the sign in page (which should not be protected)?

In conclusion, my checks to find out whether the request is for any page look like this:

```ts
// Note: I'm using a root wildcard here, as we want to protect every single page for authentication.
export const config = {
  matcher: ["/:path*"],
};

const PUBLIC_FILE_CHECK = /\.(.*)$/;

export default async function _middleware(req: NextRequest) {
  /**
   * The pathname of the route.
   * The real URL with the real dynamic part/slug value (not the slug name/ID):
   *
   * `/some-route/abc/123`
   */
  const pathname = req.nextUrl.pathname;

  /*
   * This root middleware will also intercept all other requests that are happening in the app (like external API calls
   * or static assets like the favicon) - that's why we here check whether the request was for a page and ignore all other requests.
   * The checks should be sorted by number of occurrences to benefit from short-circuiting.
   */
  if (
    // Next.js request
    !!pathname.startsWith("/_next") ||
    // API request
    !!pathname.includes("/api/") ||
    // API external request
    !!pathname.includes("/my-custom-api/") ||
    // public file request
    !!PUBLIC_FILE_CHECK.test(pathname) ||
    // Sign in page request
    !!pathname.startsWith(ROUTES.SIGN_IN)
  ) {
    // the request is not for one of our pages, go on as-is
    return NextResponse.next();
  } else {
    // this is a request for a page, do the auth check...
  }
}
```

It would be so nice if it would be able to switch all the checks around by having a way to identify just requests that are for pages.

**Comment by balazsorban44** [comment]
Hi, check out https://nextjs.org/docs/advanced-features/middleware#matcher

**Comment by bennettdams** [comment]
Hey @balazsorban44,

I saw the new matchers, but for the case of authentication (every page should only be accessible for authenticated users), I think this is a very brittle solution. If you add a new page to your app and forget to add it to the matchers, this page is accessible without authentication.

Using a root wildcard matcher (like `/:path*`) also does not work, because now **all** requests, even non-page requests, are included again.

I would really like to just protect every single page, but this is not possible anymore (I think) without `page`.

**Comment by tiepvuvan** [comment]
I'm also disappointed with the new matcher, we don't want to validate assets request like `.json`, `.js`. We only want to protect some pages.

New matcher doesn't include options like `/:teamId/:path*`, is that impossible to protect dynamic route paths?

Also, why did you close this issue without a solution?  @balazsorban44

**Reply by balazsorban44** [reply]
:+1: My bad, I meant to convert it to a discussion, as all feature requests are converted https://github.com/vercel/next.js/blob/canary/contributing.md#triaging.

**Comment by harisvsulaiman** [comment]
This broke my middleware based authentication solution for me. I think the page api should be reintroduced.

**Reply by bryanprimus** [reply]
same for me. in my case, I really need to know the value of the `request.page.name` to handle authentication for my dynamic page since it returns the name of the file we specified under the `pages` dir. for example, if we have `pages/[customerSlug]/index.tsx`, when we visit that page, the `request.page.name` will return `[customerSlug]`.

In middleware, I use this to validate whether the user is a customer or not.

```ts
const middleware = (req) => {
  const user = getSession(req)

  if(req.page.name === "[customerSlug]") {
     if(user.role === "customer") {
       return NextResponse.next()
     }
     
     return NextResponse.rewrite(to404Page)
  
  }
}
```

**Comment by neb-b** [comment]
Has anyone found a solution to this that doesn't involve checking against non-dynamic routes?

My app has many top level static routes and one dynamic route (`/download`, `/developer`, `/[username]`). There doesn't seem to be a way to handle the `/[username]` route differently without checking against all static routes.
